src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value_impl.hpp

```diff
@@ -0,0 +1,457 @@
+/**
+ * @file discrete_hilbert_value.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Defintion of the DiscreteHilbertValue class, a class that calculates
+ * the ordering of points using the Hilbert curve.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_DISCRETE_HILBERT_VALUE_IMPL_HPP
+
+#include "discrete_hilbert_value.hpp"
+
+namespace mlpack {
+namespace tree /** Trees and tree-building procedures. */ {
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue() :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(NULL),
+    ownsValueToInsert(false)
+{
+
+}
+
+template<typename TreeElemType>
+DiscreteHilbertValue<TreeElemType>::~DiscreteHilbertValue()
+{
+  if (ownsLocalDataset)
+    delete localDataset;
+  if (ownsValueToInsert)
+    delete valueToInsert;
+}
+
+template<typename TreeElemType>
+template<typename TreeType>
+DiscreteHilbertValue<TreeElemType>::DiscreteHilbertValue(const TreeType* tree) :
+    localDataset(NULL),
+    ownsLocalDataset(false),
+    numValues(0),
+    valueToInsert(tree->Parent() ? 
+                tree->Parent()->AuxiliaryInfo().HilbertValue().ValueToInsert() :
+                new arma::Col<HilbertElemType>(tree->Dataset().n_rows)),
+    ownsValueToInsert(tree->Parent() ? false : true)
+{
+  // Calculate the Hilbert value for all points
+  if (!tree->Parent()) //  This is the root node
+    ownsLocalDataset = true;
+  else if (tree->Parent()->Children()[0]->IsLeaf())
+  {
+    // This is a leaf node
+    assert(tree->Parent()->NumChildren() > 0);
+    ownsLocalDataset = true;
+  }
```

Comment:
I agree that `hilbertValues` is a better name. But I think that `localDataset` works fine and since we decided that we should add something like `InsertPoints()` into `RectangleTree` we have to update `hilbertValues` on each insertion/deletion. Right now `localDataset` does not require that. What do you think about it?